Store footprints in localStorage, various layout changes and gas consumption unit conversion

manue1 opened the pull request:
Store and read user footprint, average footprint and take action results in/from localStorage.
Update default value of all sliders to be read from user footprint (default or user-defined value).

Add gas consumption selector for switching between distance: miles & km and gas consumption: MPG  & L / 100km

Commit:
Store and read user footprint, average footprint and take action results in/from localStorage.

Update default value of all sliders to be read from user footprint (default or user-defined value).

Commit:
Show overall chart by default on desktop.

Commit:
Rearrange Footprint Summary: show User Footprint, Average Footprint, Comparison and Total in one row.

Commit:
Update slider tick style to signalize that it's clickable. Center sliders and their label.

Commit:
Add L/100km gas consumption unit conversion to Travel section.

erichulburd commented on shared/components/graphs/graphs.component.js:
```diff
@@ -215,8 +231,31 @@ class GraphsComponent extends Panel {
+      graphs.state.show_total = false;
+    } else {
+      graphs.state.show_total = true;
+    }
```
If you want to change state in react, you always have to use `component.getState` method - `state['something'] = value` won't update the DOM.

Another note, we should avoid changing state in getters. State variables should only change in React lifecycle events ([React lifecycle](https://facebook.github.io/react/docs/component-specs.html)), or UI events (`onClick`, `onChange`, etc) - or methods that those events call (you don't literally have to cram them into those methods).

manue1 replied:
This route condition was actually redundant. Anyways made sure that state is not changed in getters:

``` javascript
componentDidMount() {
    let graphs = this;
    if (window.outerWidth < 992) {
      graphs.setState({
        show_chart: false
      });
    }
```

erichulburd commented on shared/components/shopping/shopping.component.js:
```diff
@@ -144,7 +144,8 @@ class ShoppingComponent extends Panel {
-    let shopping = this;
+    let shopping = this,
+        total_services_api_key = shopping.apiKey('input_footprint_shopping_services_total');
 
```
the `apiKey` method is a helper for using api key abbreviations, so you only have to pass the part beyond what's in `api_key_base`. In shopping.component.js it's

```

  get api_key_base(){
    return 'input_footprint_shopping';
  }
```

so just `apiKey('services_total')` will work.

manue1 replied:
Within the shopping component you overwrite the apiKey footprintable method checking if the type is part of the GOODS_QUESTIONS array.
So when using this method it returned `input_footprint_shopping_services_services_total` and `input_footprint_shopping_services_goods_total`.

I have instead hard coded the keys for the total amounts within the services and goods slider init.

erichulburd commented on shared/components/travel/travel.component.js:
```diff
@@ -166,6 +173,76 @@ class TravelComponent extends Panel {
+    let travel = this;
+    if (travel.state.consumption_unit === 'mpg') return ' mi)';
+    else return ' km)';
+  }
```
We should prob translate 'mi' and 'km' here.

manue1 replied:
Abbreviations are now translated:
`"miles_abbr": "mi",   "km_abbr": "km",`

erichulburd commented on shared/lib/state_manager/state_manager.js:
```diff
@@ -99,9 +117,16 @@ export default class StateManager {
     let state_manager = this;
     // we'll load past user answers and get CC results here.
+    state_manager.checkLocalStorage();
     return state_manager.updateDefaults();
```
This should be either or. If there's data in localStorage, just load the localStorage values, no need to call api again (return `Promise.resolve()` if need be).

Otherwise, just make the api call and answers should be saved.

We may also want to add a button to clear all values from localStorage ('Forget my answers'... something like that).

manue1 replied:
setAPI is now only called if footprint hasn't already been stored in localStorage:

``` javascript
getInitialData(){
    let state_manager = this;
    // we'll load past user answers and get CC results here.
    return state_manager.checkLocalStorage();
  }

checkLocalStorage(){
    let state_manager = this;
    if (state_manager.take_action_storage) {
      state_manager.state['result_takeaction_dollars'] = JSON.parse(state_manager.take_action_storage.dollars);
      state_manager.state['result_takeaction_net10yr'] = JSON.parse(state_manager.take_action_storage.net10yr);
      state_manager.state['result_takeaction_pounds'] = JSON.parse(state_manager.take_action_storage.pounds);
    }
    if (state_manager.average_footprint_storage && state_manager.user_footprint_storage) {
      Object.assign(state_manager.state.average_footprint, state_manager.average_footprint_storage)
      Object.assign(state_manager.state.user_footprint, state_manager.user_footprint_storage)
      return Promise.resolve();
    } else {
      return state_manager.updateDefaults();
    }
  }
```

erichulburd commented on shared/lib/state_manager/state_manager.js:
```diff
@@ -99,9 +117,16 @@ export default class StateManager {
+    if (state_manager.average_footprint_storage) Object.assign(state_manager.state.average_footprint, state_manager.average_footprint_storage)
+    if (state_manager.user_footprint_storage) Object.assign(state_manager.state.user_footprint, state_manager.user_footprint_storage)
+  }
+
```
We should also save the take action values from `parseTakeactionResult`.

manue1 replied:
Take action values are now also stored in localStorage:

``` javascript
  updateTakeActionResultStorage() {
    let state_manager = this;
    localStorage.setItem('result_takeaction_dollars', JSON.stringify(state_manager.state['result_takeaction_dollars']));
    localStorage.setItem('result_takeaction_net10yr', JSON.stringify(state_manager.state['result_takeaction_net10yr']));
    localStorage.setItem('result_takeaction_pounds', JSON.stringify(state_manager.state['result_takeaction_pounds']));
  }
```

These get updated by `parseTakeactionResult`

Commit:
Review and make changes according to PR comments.

State changes no longer used in getters.
Include translation of mi / km.
UpdateDefaults only if localStorage values are not set.
Store TakeActionResults in localStorage.

Commit:
Add 'Forget my answers' button to Get Started component for resetting user-input to default values.

erichulburd committed:
convert on toggle consumption units; take action calc bug fix; 0% diff on average

erichulburd committed:
media query for graph summary figures

Commit:
Merge pull request #7 from Arbolista/feature/production_build

Feature/production build

Arbolista commented:
@manue1 awesome looks good!
